ambari-server/src/main/java/org/apache/ambari/server/api/rest/BootStrapResource.java

```diff
@@ -66,13 +67,15 @@ public static void init(BootStrapImpl instance) {
    * @throws Exception
    */
   @POST @ApiIgnore // until documented
+  @Path("/{validations}")
```

Comment:
This would mean even a call such as POST /api/v1/bootstrap/SomeRandomString would run validations which is a side effect. 

Also the general practice we have followed is to call such APIs as dry_run. For example do a dry run for deleting a host (https://issues.apache.org/jira/browse/AMBARI-16145) or do a dry run for VDF https://reviews.apache.org/r/46100/. 

I would recommend to make the API as follows

POST /api/v1/bootstrap 
{
  "verbose" : true,
  "dryRun" : true,
  "sshKey" : "sshKey",
  "user" : "root",
 "sshPort" : "22",
"userRunAs" : "root",
"hosts" : ["host1", "host2", "host3"]
}

Reply:
actually in my next patch, it will be:

/**
   * Run bootstrap on a list of hosts.
   * @response.representation.200.doc
   *
   * @response.representation.200.mediaType application/json
   * @response.representation.406.doc Error in format
   * @response.representation.408.doc Request Timed out
   * @throws Exception
   */
  @POST @ApiIgnore // until documented
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
  public BSResponse bootStrap(SshHostInfo sshInfo, @PathParam("validations") String validations,
                              @Context UriInfo uriInfo) {
    
    normalizeHosts(sshInfo);

    BSResponse resp = bsImpl.runBootStrap(sshInfo, false);

    return resp;
  }
  /**
   * Run bootstrap on a list of hosts.
   * @response.representation.200.doc
   *
   * @response.representation.200.mediaType application/json
   * @response.representation.406.doc Error in format
   * @response.representation.408.doc Request Timed out
   * @throws Exception
   */
  @POST @ApiIgnore // until documented
  @Path("/validations")
  @Consumes(MediaType.APPLICATION_JSON)
  @Produces({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
  public BSResponse bootStrap(SshHostInfo sshInfo,
                              @Context UriInfo uriInfo) {

    normalizeHosts(sshInfo);

    BSResponse resp = bsImpl.runBootStrap(sshInfo, true);

    return resp;
  }

The reason is that jax-rs seems not to support optional path param very well and it will redirect to GET api/v1/bootstrap. So I decide to seprate these two requests.

Reply:
the new patch should have removed side effect of request path.